browser/modules/PlacesProvider.jsm

```diff
@@ -0,0 +1,258 @@
+    },
+
+    QueryInterface: XPCOMUtils.generateQI([Ci.nsINavHistoryObserver,
+                        Ci.nsISupportsWeakReference])
+  },
+
+  /**
+   * Must be called before the provider is used.
+   */
+  init: function PlacesProvider_init() {
+    PlacesUtils.history.addObserver(this.historyObserver, true);
+  },
+
+  /**
+   * Gets the current set of links delivered by this provider.
+   *
+   * @returns {Promise} Returns a promise with the array of links as payload.
+   */
+  getLinks: function PlacesProvider_getLinks() {
+    let getLinksPromise = new Promise((resolve, reject) => {
+      let options = PlacesUtils.history.getNewQueryOptions();
+      options.maxResults = this.maxNumLinks;
+
+      // Sort by frecency, descending.
+      options.sortingMode = Ci.nsINavHistoryQueryOptions
+        .SORT_BY_FRECENCY_DESCENDING;
+
+      let links = [];
+
+      let callback = {
+        handleResult: function(aResultSet) {
+          let row;
+
+          while ((row = aResultSet.getNextRow())) {
+            let url = row.getResultByIndex(1);
+            if (LinkChecker.checkLoadURI(url)) {
+              let title = row.getResultByIndex(2);
+              let frecency = row.getResultByIndex(12);
+              let lastVisitDate = row.getResultByIndex(5);
+              links.push({
+                url: url,
+                title: title,
+                frecency: frecency,
+                lastVisitDate: lastVisitDate,
+                type: "history",
+              });
+            }
+          }
+        },
+
+        handleError: function(aError) {
+          reject(aError);
+        },
+
+        handleCompletion: function(aReason) { // jshint ignore:line
+          // The Places query breaks ties in frecency by place ID descending, but
+          // that's different from how Links.compareLinks breaks ties, because
+          // compareLinks doesn't have access to place IDs.  It's very important
+          // that the initial list of links is sorted in the same order imposed by
+          // compareLinks, because Links uses compareLinks to perform binary
+          // searches on the list.  So, ensure the list is so ordered.
+          let i = 1;
+          let outOfOrder = [];
+          while (i < links.length) {
+            if (LinkUtils.compareLinks(links[i - 1], links[i]) > 0) {
+              outOfOrder.push(links.splice(i, 1)[0]);
+            } else {
+              i++;
+            }
+          }
+          for (let link of outOfOrder) {
+            i = BinarySearch.insertionIndexOf(LinkUtils.compareLinks, links, link);
+            links.splice(i, 0, link);
+          }
+
+          resolve(links);
+        }
+      };
+
+      // Execute the query.
+      let query = PlacesUtils.history.getNewQuery();
+      let db = PlacesUtils.history.QueryInterface(Ci.nsPIPlacesDatabase);
+      db.asyncExecuteLegacyQueries([query], 1, options, callback);
+    });
+
+    return getLinksPromise;
+  }
+};
+
+let PlacesProvider = {
```

Comment:
The reason PlacesProvider is provided as an object is to make `XPCOMUtils.defineLazyModuleGetter` easier.

A user of `PlacesProvider.jsm` would otherwise need more lines of code.

Instead of:
```javascript
XPCOMUtils.defineLazyModuleGetter(this, "PlacesProvider",
  "resource://gre/modules/PlacesProvider.jsm");
```

They'd need to write:
```javascript
XPCOMUtils.defineLazyModuleGetter(this, "Links", function() {
  const {Links} Cu.import("resource://gre/modules/PlacesProvider.jsm");
  return Links;
});
```